File: WHC_KeyboradManager(OC)/WHC_KeyboradManager(OC)/WHC_KeyboradManager(OC)/WHC_KeyboardHeaderView.m
//
//  WHC_KeyboardHeaderView.m
//  WHC_KeyboardManager(OC)
//
//  Created by WHC on 16/11/20.
//  Copyright © 2016年 WHC. All rights reserved.
//
//  Github <https://github.com/netyouli/WHC_KeyboardManager>

//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "WHC_KeyboardHeaderView.h"
#import "WHC_KeyboardManager.h"

@implementation WHC_KeyboardHeaderView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor whiteColor];
        if ([UIDevice currentDevice].systemVersion.floatValue >= 8.0) {
            UIBlurEffect * blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleExtraLight];
            UIVisualEffectView * blurEffectView = [[UIVisualEffectView alloc] initWithEffect:blurEffect];
            [self addSubview:blurEffectView];
            blurEffectView.translatesAutoresizingMaskIntoConstraints = NO;
            [self addConstraint:[NSLayoutConstraint constraintWithItem:blurEffectView attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeLeft multiplier:1 constant:0]];
            [self addConstraint:[NSLayoutConstraint constraintWithItem:blurEffectView attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeRight multiplier:1 constant:0]];
            [self addConstraint:[NSLayoutConstraint constraintWithItem:blurEffectView attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1 constant:0]];
            [self addConstraint:[NSLayoutConstraint constraintWithItem:blurEffectView attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeBottom multiplier:1 constant:0]];
        }
        
        _nextButton = [UIButton new];
        _frontButton = [UIButton new];
        _doneButton = [UIButton new];
        _lineView = [UIView new];
        
        [self addSubview:_nextButton];
        [self addSubview:_frontButton];
        [self addSubview:_doneButton];
        [self addSubview:_lineView];
        
        _nextButton.translatesAutoresizingMaskIntoConstraints = NO;
        _frontButton.translatesAutoresizingMaskIntoConstraints = NO;
        _doneButton.translatesAutoresizingMaskIntoConstraints = NO;
        _lineView.translatesAutoresizingMaskIntoConstraints = NO;
        
        _lineView.backgroundColor = [UIColor colorWithWhite:0.8 alpha:1];
        [_frontButton setTitle:@"←" forState: UIControlStateNormal];
        [_nextButton setTitle:@"→" forState: UIControlStateNormal];
        [_doneButton setTitle:@"完成" forState: UIControlStateNormal];

        [_frontButton setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        [_nextButton setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        [_doneButton setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        
        [_frontButton setTitleColor:[UIColor grayColor] forState:UIControlStateSelected];
        [_nextButton setTitleColor:[UIColor grayColor] forState:UIControlStateSelected];
        
        _frontButton.titleLabel.font = [UIFont boldSystemFontOfSize:20];
        _nextButton.titleLabel.font = [UIFont boldSystemFontOfSize:20];

        [_frontButton addTarget:self action:@selector(clickFront:) forControlEvents:UIControlEventTouchUpInside];
        [_nextButton addTarget:self action:@selector(clickNext:) forControlEvents:UIControlEventTouchUpInside];
        [_doneButton addTarget:self action:@selector(clickDone:) forControlEvents:UIControlEventTouchUpInside];
        
        CGFloat kMargin = 0;
        CGFloat kWidth = 60;
        
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_frontButton attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeLeft multiplier:1 constant:kMargin]];
        [_frontButton addConstraint:[NSLayoutConstraint constraintWithItem:_frontButton attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:kWidth]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_frontButton attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeBottom multiplier:1 constant:0]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_frontButton attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1 constant:0]];
        
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_nextButton attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:_frontButton attribute:NSLayoutAttributeRight multiplier:1 constant:kMargin]];
        [_nextButton addConstraint:[NSLayoutConstraint constraintWithItem:_nextButton attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:kWidth]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_nextButton attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeBottom multiplier:1 constant:0]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_nextButton attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1 constant:0]];

        [self addConstraint:[NSLayoutConstraint constraintWithItem:_doneButton attribute:NSLayoutAttributeTrailing relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTrailing multiplier:1 constant:-kMargin]];
        [_doneButton addConstraint:[NSLayoutConstraint constraintWithItem:_doneButton attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:kWidth]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_doneButton attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeBottom multiplier:1 constant:0]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_doneButton attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1 constant:0]];

        [self addConstraint:[NSLayoutConstraint constraintWithItem:_lineView attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeLeft multiplier:1 constant:0]];
        [_lineView addConstraint:[NSLayoutConstraint constraintWithItem:_lineView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:0.5]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_lineView attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeRight multiplier:1 constant:0]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_lineView attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1 constant:0]];
        
        /// 监听WHC_KeyboardManager通知
        NSNotificationCenter * nCenter = [NSNotificationCenter defaultCenter];
        [nCenter addObserver:self selector:@selector(getCurrentFieldView:) name:(NSString *)WHC_KBM_CurrentFieldView object:nil];
        [nCenter addObserver:self selector:@selector(getNextFieldView:) name:(NSString *)WHC_KBM_NextFieldView object:nil];
        [nCenter addObserver:self selector:@selector(getFrontFieldView:) name:(NSString *)WHC_KBM_FrontFieldView object:nil];
        
    }
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - 获取WHC_KeyboardManager发来的通知 -
- (void)getCurrentFieldView:(NSNotification *)notify {
    _currentFieldView = notify.object;
}

- (void)getFrontFieldView:(NSNotification *)notify {
    _frontFieldView = notify.object;
    _frontButton.selected = _frontFieldView == nil;
}

- (void)getNextFieldView:(NSNotification *)notify {
    _nextFieldView = notify.object;
    _nextButton.selected = _nextFieldView == nil;
}

#pragma mark - ACTION -
- (void)clickFront:(UIButton *)button {
    if (_frontFieldView) {
        if ([_frontFieldView isKindOfClass:[UITextField class]]) {
            [((UITextField *)_frontFieldView) becomeFirstResponder];
        }else if ([_frontFieldView isKindOfClass:[UITextView class]]) {
            [((UITextView *)_frontFieldView) becomeFirstResponder];
        }
    }
    if (_clickFrontButtonBlock) {
        _clickFrontButtonBlock();
    }
}

- (void)clickNext:(UIButton *)button {
    if (_nextFieldView) {
        if ([_nextFieldView isKindOfClass:[UITextField class]]) {
            [((UITextField *)_nextFieldView) becomeFirstResponder];
        }else if ([_nextFieldView isKindOfClass:[UITextView class]]) {
            [((UITextView *)_nextFieldView) becomeFirstResponder];
        }
    }
    if (_clickNextButtonBlock) {
        _clickNextButtonBlock();
    }
}

- (void)clickDone:(UIButton *)button {
    if (_currentFieldView) {
        if ([_currentFieldView isKindOfClass:[UITextField class]]) {
            [((UITextField *)_currentFieldView) resignFirstResponder];
        }else if ([_currentFieldView isKindOfClass:[UITextView class]]) {
            [((UITextView *)_currentFieldView) resignFirstResponder];
        }
    }
    if (_clickDoneButtonBlock) {
        _clickDoneButtonBlock();
    }
}

@end

File: WHC_KeyboradManager/WHC_KeyboradManager/WHC_KeyboardHeaderView.swift
//
//  WHC_KeyboardHeaderView.swift
//  WHC_KeyboardManager
//
//  Created by WHC on 16/11/16.
//  Copyright © 2016年 WHC. All rights reserved.
//

//  Github <https://github.com/netyouli/WHC_KeyboardManager>

//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import UIKit

extension UIColor {
    static var kbBlack: UIColor {
        if #available(iOS 13.0, *) {
            return UIColor { (trait) -> UIColor in
                if trait.userInterfaceStyle == .dark {
                    return UIColor(white: 1, alpha: 1)
                }
                return UIColor(white: 0, alpha: 1)
            }
        } else {
            // Fallback on earlier versions
            return UIColor.black
        }
    }
    
    static var kbLine: UIColor {
        if #available(iOS 13.0, *) {
            return UIColor { (trait) -> UIColor in
                if trait.userInterfaceStyle == .dark {
                    return UIColor(white: 0.2, alpha: 1)
                }
                return UIColor(red: 244.0 / 255.0, green: 244.0 / 255.0, blue: 244.0 / 255.0, alpha: 1)
            }
        } else {
            // Fallback on earlier versions
            return UIColor(red: 244.0 / 255.0, green: 244.0 / 255.0, blue: 244.0 / 255.0, alpha: 1)
        }
    }
}

public class WHC_KeyboardHeaderView: UIView {

    private(set) public var currentFieldView: UIView?
    private(set) public var nextFieldView: UIView?
    private(set) public var frontFieldView: UIView?
    
    private let kMargin: CGFloat = 0
    private let kWidth: CGFloat = 60
    
    /// 点击前一个按钮回调
    public var clickFrontButtonBlock: (() -> Void)!
    /// 点击下一个按钮回调
    public var clickNextButtonBlock: (() -> Void)!
    /// 点击完成按钮回调
    public var clickDoneButtonBlock: (() -> Void)!
    
    /// 只读
    private(set) public lazy var nextButton = UIButton()
    private(set) public lazy var frontButton = UIButton()
    private(set) public lazy var doneButton = UIButton()
    private(set) public lazy var lineView = UIView()
    private lazy var toolBar = UIToolbar(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 44))
    
    /// 隐藏上一个下一个按钮只保留done按钮
    public var hideNextAndFrontButton = false {
        didSet {
            frontButton.isHidden = hideNextAndFrontButton
            nextButton.isHidden = hideNextAndFrontButton
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.clear
        toolBar.backgroundColor = UIColor.clear
        self.addSubview(toolBar)
        self.addSubview(nextButton)
        self.addSubview(frontButton)
        self.addSubview(doneButton)
        self.addSubview(lineView)
        
        lineView.backgroundColor = UIColor.kbLine
        
        frontButton.setTitle("←", for: .normal)
        nextButton.setTitle("→", for: .normal)
        doneButton.setTitle("完成", for: .normal)
        
        frontButton.setTitleColor(UIColor.kbBlack, for: .normal)
        nextButton.setTitleColor(UIColor.kbBlack, for: .normal)
        doneButton.setTitleColor(UIColor.kbBlack, for: .normal)
        
        frontButton.setTitleColor(UIColor.gray, for: .selected)
        nextButton.setTitleColor(UIColor.gray, for: .selected)
        
        frontButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 20)
        nextButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 20)
        
        frontButton.addTarget(self, action: #selector(clickFront(button:)), for: .touchUpInside)
        nextButton.addTarget(self, action: #selector(clickNext(button:)), for: .touchUpInside)
        doneButton.addTarget(self, action: #selector(clickDone(button:)), for: .touchUpInside)
    
        frontButton.translatesAutoresizingMaskIntoConstraints = false
        nextButton.translatesAutoresizingMaskIntoConstraints = false
        doneButton.translatesAutoresizingMaskIntoConstraints = false
        lineView.translatesAutoresizingMaskIntoConstraints = false
        toolBar.translatesAutoresizingMaskIntoConstraints = false
        
        self.addConstraint(NSLayoutConstraint(item: toolBar, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint(item: toolBar, attribute: .right, relatedBy: .equal, toItem: self, attribute: .right, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint(item: toolBar, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint(item: toolBar, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0))
        
        
        
        self.addConstraint(NSLayoutConstraint(item: frontButton, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: kMargin))
        
        frontButton.addConstraint(NSLayoutConstraint(item: frontButton, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1, constant: kWidth))
        
        self.addConstraint(NSLayoutConstraint(item: frontButton, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: frontButton, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: nextButton, attribute: .left, relatedBy: .equal, toItem: frontButton, attribute: .right, multiplier: 1, constant: kMargin))
        
        nextButton.addConstraint(NSLayoutConstraint(item: nextButton, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1, constant: kWidth))
        
        self.addConstraint(NSLayoutConstraint(item: nextButton, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: nextButton, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: doneButton, attribute: .trailing, relatedBy: .equal, toItem: self, attribute: .trailing, multiplier: 1, constant: -kMargin))
        
        doneButton.addConstraint(NSLayoutConstraint(item: doneButton, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1, constant: kWidth))
        
        self.addConstraint(NSLayoutConstraint(item: doneButton, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: doneButton, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1, constant: 0))
    
        
        self.addConstraint(NSLayoutConstraint(item: lineView, attribute: .left, relatedBy: .equal, toItem: self, attribute: .left, multiplier: 1, constant: 0))
        
        lineView.addConstraint(NSLayoutConstraint(item: lineView, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1, constant: 0.5))
        
        self.addConstraint(NSLayoutConstraint(item: lineView, attribute: .right, relatedBy: .equal, toItem: self, attribute: .right, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: lineView, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1, constant: 0))
        
        /// 监听WHC_KeyboardManager通知
        NotificationCenter.default.addObserver(self, selector: #selector(getCurrentFieldView(notify:)), name: NSNotification.Name.CurrentFieldView, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(getFrontFieldView(notify:)), name: NSNotification.Name.FrontFieldView, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(getNextFieldView(notify:)), name: NSNotification.Name.NextFieldView, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    //MARK: - 获取WHC_KeyboardManager发来的通知 -
    @objc private func getCurrentFieldView(notify: Notification) {
        currentFieldView = notify.object as? UIView
    }
    
    @objc private func getFrontFieldView(notify: Notification) {
        frontFieldView = notify.object as? UIView
        frontButton.isSelected = frontFieldView == nil
    }
    
    @objc private func getNextFieldView(notify: Notification) {
        nextFieldView = notify.object as? UIView
        nextButton.isSelected = nextFieldView == nil
    }
    
    //MARK: - ACTION -
    @objc private func clickFront(button: UIButton) {
        if WHC_KeyboardManager.share.moveDidAnimation {
            if frontFieldView != nil {
                if frontFieldView is UITextField {
                    (frontFieldView as? UITextField)?.becomeFirstResponder()
                }else if frontFieldView is UITextView {
                    (frontFieldView as? UITextView)?.becomeFirstResponder()
                }
            }
            clickFrontButtonBlock?()
        }
    }
    
    @objc private func clickNext(button: UIButton) {
        if WHC_KeyboardManager.share.moveDidAnimation {
            if nextFieldView != nil {
                if nextFieldView is UITextField {
                    (nextFieldView as? UITextField)?.becomeFirstResponder()
                }else if nextFieldView is UITextView {
                    (nextFieldView as? UITextView)?.becomeFirstResponder()
                }
            }
            clickNextButtonBlock?()
        }
    }
    
    @objc private func clickDone(button: UIButton) {
        if currentFieldView != nil {
            if currentFieldView is UITextField {
                currentFieldView?.resignFirstResponder()
            }else if currentFieldView is UITextView {
                currentFieldView?.resignFirstResponder()
            }
        }
        clickDoneButtonBlock?()
    }

}
